plugin/kubernetescrd/controller.go

```diff
@@ -0,0 +1,235 @@
+				}
+				return controller.client.Resource(corednsv1alpha1.GroupVersion.WithResource("dnszones")).Watch(ctx, options)
+			},
+		},
+		&unstructured.Unstructured{},
+		defaultResyncPeriod,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				key, err := cache.MetaNamespaceKeyFunc(obj)
+				if err == nil {
+					controller.workqueue.Add(key)
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				key, err := cache.MetaNamespaceKeyFunc(newObj)
+				if err == nil {
+					controller.workqueue.Add(key)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+				if err == nil {
+					controller.workqueue.Add(key)
+				}
+			},
+		},
+	)
+
+	return &controller
+}
+
+// Run starts the controller. Threads is the number of workers that can process
+// work on the workqueue in parallel.
+func (d *dnsZoneCRDControl) Run(threads int) {
+	defer utilruntime.HandleCrash()
+	defer d.workqueue.ShutDown()
+
+	go d.dnsZoneController.Run(d.stopCh)
+
+	if !cache.WaitForCacheSync(d.stopCh, d.dnsZoneController.HasSynced) {
+		utilruntime.HandleError(errors.New("Timed out waiting for caches to sync"))
+		return
+	}
+
+	for i := 0; i < threads; i++ {
+		go wait.Until(d.runWorker, time.Second, d.stopCh)
+	}
+
+	<-d.stopCh
+
+	// Shutdown all plugins
+	for _, plugin := range d.pluginMap.List() {
+		plugin.OnShutdown()
+	}
+}
+
+// HasSynced returns true once the controller has completed an initial resource
+// listing.
+func (d *dnsZoneCRDControl) HasSynced() bool {
+	return d.dnsZoneController.HasSynced()
+}
+
+// Stop stops the controller.
+func (d *dnsZoneCRDControl) Stop() error {
+	d.stopLock.Lock()
+	defer d.stopLock.Unlock()
+
+	// Only try draining the workqueue if we haven't already.
+	if !d.shutdown {
+		close(d.stopCh)
+		d.shutdown = true
+
+		return nil
+	}
+
+	return fmt.Errorf("shutdown already in progress")
+}
+
+func (d *dnsZoneCRDControl) runWorker() {
+	for d.processNextItem() {
+	}
+}
+
+func (d *dnsZoneCRDControl) processNextItem() bool {
+	key, quit := d.workqueue.Get()
+	if quit {
+		return false
+	}
+
+	defer d.workqueue.Done(key)
+
+	err := d.sync(key.(string))
+	if err != nil {
+		log.Errorf("Error syncing DNSZone %v: %v", key, err)
+		d.workqueue.AddRateLimited(key)
+		return true
+	}
+
+	d.workqueue.Forget(key)
+
+	return true
+}
+
+func (d *dnsZoneCRDControl) sync(key string) error {
+	obj, exists, err := d.dnsZoneLister.GetByKey(key)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		plugin := d.pluginMap.Delete(key)
+		if plugin != nil {
+			plugin.OnShutdown()
+		}
+	} else {
+		dnsZone, err := d.convertToDNSZone(obj.(runtime.Object))
+		if err != nil {
+			return err
+		}
+		forwardConfig := forward.ForwardConfig{
+			From: dnsZone.Spec.ZoneName,
+			To:   []string{dnsZone.Spec.ForwardTo},
+		}
+		plugin, err := d.instancer(forwardConfig)
+		if err != nil {
+			return err
+		}
+		err = plugin.OnStartup()
```

Comment:
This doesnt capture the dnstap integration (from forward setup).  Not sure if it's achievable from this point.

Reply:
Good catch, but what makes you say you don't think it is achievable? I think I can save off the `dnstap` plugin within setup and set the `tapPlugin` on `forward.NewWithConfig`, but let me know if there is something I'm missing.

Reply:
Oh it could be - I didn't scratch below the surface ...  just saying I wasn't sure if it was or not.

Reply:
Gotcha, I implemented dnstap and tested it locally.